---
title: "01-HOPE-HGG-subtyping-subset-file"
output: html_document
date: "2023-08-02"
---

```{r library}
library(tidyverse)
library(annoFuseData)

```

## Get directories

```{r directories}
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
scratch_dir <- file.path(root_dir, "scratch")
analyses_dir <- hgg_subset_dir <- file.path(root_dir, 
                        "analyses",
                        "molecular-subtyping-HGG")

# Get hgg subset folder from the module in master
hgg_subset_dir <- file.path(analyses_dir,
                        "hgg-subset")
# File path to results directory
results_dir <-
  file.path(analyses_dir, "results")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}



```

## load histologies files

```{r pressure, echo=FALSE}
hist <- readr::read_tsv(file.path(root_dir, "data", "Hope-GBM-histologies-base.tsv"))

```

# Prepare Data
## Defining lesions snv data 

```{r}
# select tumor sample barcode, gene, short protein annotation and variant classification
keep_cols <- c("Chromosome",
             "Start_Position",
             "End_Position",
             "Strand",
             "Variant_Classification",
             "IMPACT",
             "Tumor_Sample_Barcode",
             "Hugo_Symbol",
             "HGVSp_Short",
             "Exon_Number")
# snv files
snv_tumor_maf <- data.table::fread(
  file.path(root_dir, "data" , "Hope-tumor-only-snv-mutect2.maf.tsv.gz"), 
  tmpdir = scratch_dir,
  select = keep_cols)

snv_consensus_hotspot_maf <- data.table::fread(
  file.path(root_dir, "data" , "Hope-snv-consensus-plus-hotspots.maf.tsv.gz"), 
  tmpdir = scratch_dir,
  select = keep_cols) %>%
  bind_rows(snv_tumor_maf)

dna_bsids <- hist %>%
  filter(!is.na(pathology_diagnosis),
                (experimental_strategy %in% c("WGS", "WXS", "Targeted Sequencing") & is.na(RNA_library))) %>%
  pull(Kids_First_Biospecimen_ID) %>% 
  unique()
```

### Identify the target lesions in Hope cohort

```{r}
# Filter the snv consensus mutation data for the target lesions
snv_lesions_df <- snv_consensus_hotspot_maf  %>%
  dplyr::filter(Hugo_Symbol %in% c("H3-3A", "H3C2",
                                   "H3C3", "H3C14") &
                  HGVSp_Short %in% c("p.K28M", "p.G35R",
                                     "p.G35V", "p.K28I")) %>%
  dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol, HGVSp_Short) %>%
  dplyr::mutate(
    `H3F3A_H3-3A.K28M` = dplyr::case_when(Hugo_Symbol == "H3-3A" &
                                    HGVSp_Short == "p.K28M" ~ "Yes",
                                  TRUE ~ "No"),
    HIST1H3B_H3C2.K28M = dplyr::case_when(
      Hugo_Symbol == "H3C2" & HGVSp_Short == "p.K28M" ~ "Yes",
      TRUE ~ "No"),
    HIST1H3C_H3C3.K28M = dplyr::case_when(
      Hugo_Symbol == "H3C3" & HGVSp_Short == "p.K28M" ~ "Yes",
      TRUE ~ "No"),
    HIST2H3C_H3C14.K28M = dplyr::case_when(
      Hugo_Symbol == "H3C14" & HGVSp_Short == "p.K28M" ~ "Yes",
      TRUE ~ "No"),
    ## add H3 p.K28I mutation
    `H3F3A_H3-3A.K28I` = dplyr::case_when(Hugo_Symbol == "H3-3A" &
                                    HGVSp_Short == "p.K28I" ~ "Yes",
                                  TRUE ~ "No"),
    HIST1H3B_H3C2.K28I = dplyr::case_when(
      Hugo_Symbol == "H3C2" & HGVSp_Short == "p.K28I" ~ "Yes",
      TRUE ~ "No"),
    HIST1H3C_H3C3.K28I = dplyr::case_when(
      Hugo_Symbol == "H3C3" & HGVSp_Short == "p.K28I" ~ "Yes",
      TRUE ~ "No"),
    HIST2H3C_H3C14.K28I = dplyr::case_when(
      Hugo_Symbol == "H3C14" & HGVSp_Short == "p.K28I" ~ "Yes",
      TRUE ~ "No"),
    `H3F3A_H3-3A.G35R` = dplyr::case_when(Hugo_Symbol == "H3-3A" &
                                    HGVSp_Short == "p.G35R" ~ "Yes",
                                  TRUE ~ "No"),
    `H3F3A_H3-3A.G35V` = dplyr::case_when(Hugo_Symbol == "H3-3A" &
                                    HGVSp_Short == "p.G35V" ~ "Yes",
                                  TRUE ~ "No")
  ) %>%
  dplyr::select(
    -HGVSp_Short,
    -Hugo_Symbol
  ) 

snv_lesions_df <- hist %>%
  dplyr::select(Kids_First_Participant_ID, 
                sample_id,
                Kids_First_Biospecimen_ID) %>%
  dplyr::full_join(snv_lesions_df,
                    by = c("Kids_First_Biospecimen_ID" = "Tumor_Sample_Barcode")) %>%
  filter(Kids_First_Biospecimen_ID %in% dna_bsids) %>%
  dplyr::mutate_all(function(x) tidyr::replace_na(x, "No")) %>%
  dplyr::mutate(
    defining_lesion = dplyr::case_when(
      `H3F3A_H3-3A.K28M` == "Yes" ~ TRUE,
      HIST1H3B_H3C2.K28M == "Yes" ~ TRUE,
      HIST1H3C_H3C3.K28M == "Yes" ~ TRUE,
      HIST2H3C_H3C14.K28M == "Yes" ~ TRUE,
      ## for K28I mutation
      `H3F3A_H3-3A.K28I` == "Yes" ~ TRUE,
      HIST1H3B_H3C2.K28I == "Yes" ~ TRUE,
      HIST1H3C_H3C3.K28I == "Yes" ~ TRUE,
      HIST2H3C_H3C14.K28I == "Yes" ~ TRUE,
      `H3F3A_H3-3A.G35R` == "Yes" ~ TRUE,
      `H3F3A_H3-3A.G35V` == "Yes" ~ TRUE,
      TRUE ~ FALSE
    )
  ) %>%
  dplyr::arrange(Kids_First_Participant_ID, sample_id) %>% 
  write_tsv(file.path(results_dir, "Hope_HGG_defining_lesions.tsv"))


```

## cn data

```{r}
cn_tumor_only_df <- readr::read_rds(file.path(root_dir, "data", 
                                   "Hope-cnv-controlfreec-tumor-only.rds"))

cn_df <- readr::read_rds(file.path(root_dir, "data", 
                                   "Hope-cnv-controlfreec.rds")) %>%
  bind_rows(cn_tumor_only_df) 

cn_status_df <- cn_df %>% 
    dplyr::filter(gene_symbol %in% c("PDGFRA", "PTEN", "MYCN", "EGFR")) %>% 
  select(gene_symbol,
         Kids_First_Biospecimen_ID,
         status) %>% 
  distinct() %>%
  group_by(Kids_First_Biospecimen_ID, gene_symbol) %>%
  summarise(full_status = paste(status, collapse = ", ")) %>% 
  spread(gene_symbol, full_status) %>%
  dplyr::rename(EGFR_focal_status = EGFR, 
                PTEN_focal_status = PTEN, 
                PDGFRA_focal_status = PDGFRA,
                MYCN_focal_status = MYCN) %>%
  full_join(hist, by = c("Kids_First_Biospecimen_ID" = "Kids_First_Biospecimen_ID")) %>%
  filter(experimental_strategy == "WGS") %>%
  select(sample_id,
         Kids_First_Biospecimen_ID,
         Kids_First_Participant_ID,
         EGFR_focal_status,
         PTEN_focal_status, 
         PDGFRA_focal_status, 
         MYCN_focal_status) %>%
  replace(is.na(.), "Neutral") %>%
  distinct() %>% 
  write_tsv(file.path(results_dir, "HGG_cleaned_cnv.tsv"))


```

## mutations

```{r}
gencode_cds_bed <- readr::read_tsv(file.path(root_dir, "scratch", "gencode.v39.primary_assembly.annotation.bed"),
                                   col_names = FALSE)
defining_lesions_df <- readr::read_tsv(file.path(results_dir, "Hope_HGG_defining_lesions.tsv")) %>% 
  rename(Tumor_Sample_Barcode = Kids_First_Biospecimen_ID)

snv_df <- snv_consensus_hotspot_maf

source(file.path(root_dir, "utils",
                 "tmb_functions.R"))


```

### H3 K28 mutant
### H3 G35 mutant
### IDH mutant
### H3.3 and IDH wildtype
### 1p/19q co-deleted oligodendrogliomas

```{r}
k28_genes_of_interest <- c("ACRV1", "TP53", "ATRX", "FGFR1", "IDH1", "BRAF")
g35_genes_of_interest <- c("ATRX", "DAXX", "TP53", "SETD2", "IDH1")
idh_genes_of_interest <- c("IDH1", "TP53")
wildtype_genes_of_interest <- c("H3-3A", "H3C2", "H3C3", "H3C14",
                                "IDH1", "TP53", "TERT")
codeleted_genes_of_interest <- c("IDH1", "TERT", "CIC", "FUBP1")


# generate a vection of all genes of interestes to subset snv file
genes_of_interest <- unique(c(k28_genes_of_interest,
                              g35_genes_of_interest,
                              idh_genes_of_interest,
                              wildtype_genes_of_interest,
                              codeleted_genes_of_interest))

```

### select the gene of interests and exclude TERT for now

```{r}
filtered_snv_df <- snv_df %>%
  dplyr::filter(Hugo_Symbol %in% genes_of_interest)

## For all genes other than _TERT_, we'll restrict to the coding sequences and drop silent mutations.
coding_summary_df <- snv_ranges_filter(maf_df = filtered_snv_df,
                                   keep_ranges = gencode_cds_bed) %>%
  dplyr::filter(Variant_Classification != "Silent",
                Hugo_Symbol != "TERT") %>% 
  dplyr::select(Tumor_Sample_Barcode,
                Hugo_Symbol) %>%
  # For each tumor biospecimen, get a comma separated
  dplyr::group_by(Tumor_Sample_Barcode) %>%
  dplyr::summarize(relevant_coding_mutations = paste(sort(unique(Hugo_Symbol)),
                                                     collapse = ", ")) %>% 
  dplyr::bind_rows(data.frame(
    Tumor_Sample_Barcode = setdiff(dna_bsids,
                                   .$Tumor_Sample_Barcode)
  )) %>%
  replace(is.na(.), "None")

```


#### TERT mutation

```{r}
tert_summary_df <- filtered_snv_df %>%
  dplyr::filter(Hugo_Symbol == "TERT",
                Variant_Classification == "5'Flank",
                Start_Position %in% c("1295113","1295135"),
                End_Position %in% c("1295113","1295135")) %>%
  dplyr::select(Tumor_Sample_Barcode,
                Variant_Classification) %>%
  dplyr::group_by(Tumor_Sample_Barcode) %>%
  dplyr::summarize(TERT_variant_classification = 
                     paste(sort(unique(Variant_Classification)),
                           collapse = ", ")) %>% 
  dplyr::bind_rows(data.frame(
    Tumor_Sample_Barcode = setdiff(dna_bsids,
                                   .$Tumor_Sample_Barcode)
  )) %>%
  replace(is.na(.), "None")

```

##### IDH mutation

```{r}
idh_summary_df <- filtered_snv_df %>%
  dplyr::filter(Hugo_Symbol == "IDH1",
                grepl("R132|R172", HGVSp_Short)) %>%
  dplyr::select(Tumor_Sample_Barcode, HGVSp_Short) %>%
  dplyr::rename(IDH1_mutation = HGVSp_Short) %>%
  dplyr::bind_rows(data.frame(
    Tumor_Sample_Barcode = setdiff(dna_bsids,
                                   .$Tumor_Sample_Barcode)
  )) %>%
  replace(is.na(.), "No R132 or R172") %>%
  distinct()
```

#### BRAF mutations

```{r}
braf_summary_df <- filtered_snv_df %>%
  dplyr::filter(Hugo_Symbol == "BRAF",
                grepl("V600E", HGVSp_Short)) %>%
  dplyr::select(Tumor_Sample_Barcode, HGVSp_Short) %>%
  dplyr::rename(BRAF_V600E = HGVSp_Short) %>%
  dplyr::bind_rows(data.frame(
    Tumor_Sample_Barcode = setdiff(dna_bsids,
                                   .$Tumor_Sample_Barcode)
  )) %>%
  replace(is.na(.), "No V600E")

```

# Combine all together and save

```{r}
final_mutations_df <- list(defining_lesions_df,
                           coding_summary_df,
                           idh_summary_df,
                           tert_summary_df,
                           braf_summary_df) %>%
  purrr::reduce(dplyr::inner_join, by = "Tumor_Sample_Barcode") %>%
  dplyr::select(-defining_lesion) %>%
  dplyr::arrange(Kids_First_Participant_ID, sample_id) %>%
  unique() %>%
  write_tsv(file.path(results_dir, "HGG_cleaned_mutation.tsv"))


```
## fusion subsetting

```{r}
fusion_hgg <- read_tsv(file.path(analyses_dir, "input", "fusion_summary_hgg_foi.tsv")) 

```


```{r}
RTK_list <- c("FGFR1", "NTRK", "MET", "ROS1", "ALK")

summary_df <- fusion_hgg %>%
  mutate(HGG_Fusion_evidence = apply(.[,colnames(.)[grepl(paste(RTK_list, collapse = "|"), colnames(.))]], 1, FUN =  function(x) {paste(names(x)[x == 1],collapse="/")})) %>%
  mutate(HGG_Fusion_counts = apply(.[,colnames(.)[grepl(paste(RTK_list, collapse = "|"), colnames(.))]], 1, FUN =  function(x) {length(names(x)[x == 1])})) %>%
  mutate(NTRK_fusions = apply(.[,colnames(.)[grepl("NTRK", colnames(.))]], 1, FUN =  function(x) {paste(names(x)[x == 1],collapse="/")})) %>% 
  mutate(MET_fusions = apply(.[,colnames(.)[grepl("MET", colnames(.))]], 1, FUN =  function(x) {paste(names(x)[x == 1],collapse="/")})) %>% 
  mutate(ROS1_fusions = apply(.[,colnames(.)[grepl("ROS1", colnames(.))]], 1, FUN =  function(x) {paste(names(x)[x == 1],collapse="/")})) %>% 
  mutate(ALK_fusions = apply(.[,colnames(.)[grepl("ALK", colnames(.))]], 1, FUN =  function(x) {paste(names(x)[x == 1],collapse="/")})) %>% 
  mutate(FGFR1_fusions = apply(.[,colnames(.)[grepl("FGFR1", colnames(.))]], 1, FUN =  function(x) {paste(names(x)[x == 1],collapse="/")})) %>% 
  select(c("Kids_First_Biospecimen_ID", "NTRK_fusions", "MET_fusions", "ROS1_fusions", "ALK_fusions", "FGFR1_fusions", "HGG_Fusion_evidence", "HGG_Fusion_counts")) %>% 
  dplyr::arrange(Kids_First_Biospecimen_ID) 

summary_df[summary_df == ""] = "None"

write_tsv(summary_df, file.path(results_dir, "HGG_cleaned_fusion.tsv"))

```

## gene expression


### Prepara gene expression data (RNA-seq) for Hope 

```{r}
filter_process_expression <- function(expression_mat) {
  # This function takes the collapsed FPKM expression matrix, selects relevant
  # columns (samples) via the Kids_First_Biospecimen_ID identifier, and then
  # log2(x + 1) transforms and z-scores the filtered matrix gene-wise.
  # It returns the z-scored matrix, where the columns are genes and the rows
  # are samples (biospecimen ID are the rownames).
  #
  # Only intended for use in the context of this script!

  # Filter to HGG samples only -- we can use hgg_metadata_df because it is
  # subset to RNA-seq samples
  filtered_expression <- expression_mat %>%
    select(intersect(hist$Kids_First_Biospecimen_ID,
                     colnames(expression_mat)))

  # Log2 transformation
  log_expression <- log2(filtered_expression + 1)

  # Scale does column centering, so we transpose first
  long_zscored_expression <- scale(t(log_expression),
                                   center = TRUE,
                                   scale = TRUE)
  return(long_zscored_expression)
}

rna_expression <- read_rds(
  file.path(root_dir, 
            "data",
            "Hope-and-CPTAC-GBM-gene-expression-rsem-tpm-collapsed.rds"))

expression_matrix <- filter_process_expression(rna_expression)

```

### Clean and Wrangle Expression Data
```{r}
clean_wrangle_expression <-
  function(filtered_expression, output_file_path) {
    # Given an expression matrix filtered to include only samples that meet
    # the HGG classification criteria, select the genes of interest. It displays
    # and writes the cleaned and filtered expression table to the given output
    # file path.
    
    cleaned_expression <- filtered_expression %>%
      as.data.frame() %>%
      rownames_to_column("Kids_First_Biospecimen_ID") %>%
      dplyr::arrange(Kids_First_Biospecimen_ID) %>%
      dplyr::select(Kids_First_Biospecimen_ID,
             FOXG1_expression_zscore = FOXG1,
             OLIG2_expression_zscore = OLIG2,
             TP73_AS1_expression_zscore = `TP73-AS1`,
             EGFR_expression_zscore = EGFR,
             EZHIP_expression_zscore = EZHIP) %>%
      distinct()
    
    # Save expression data to file
    write_tsv(cleaned_expression, output_file_path)
    
    # Display `cleaned_expression`
    cleaned_expression
  }
# Run `clean_wrangle_expression` on filtered expression data
cleaned_expression <-
  clean_wrangle_expression(expression_matrix, 
                           output_file_path = file.path(results_dir, "HGG_cleaned_expression.tsv"))

```


```{r}
sessionInfo()

```

